src/lib.rs

```diff
@@ -0,0 +1,433 @@
+            .number(stream_id)
+            .address(&stream.to)
+            .number(stream.amount)
+            .notify();
+        return true;
+    }
+    false
+}
+
+fn cancel_stream(stream_id: U128) -> bool {
+    if let Some(stream) = get_stream(stream_id) {
+        assert!(runtime::check_witness(&stream.to) || runtime::check_witness(&stream.from));
+        let self_addr = runtime::address();
+        let from_balance = balance_of(stream_id, &stream.from);
+        let to_balance = balance_of(stream_id, &stream.from);
+        assert!(transfer(
+            &stream.token,
+            &self_addr,
+            &stream.from,
+            from_balance
+        ));
+        assert!(transfer(
+            &stream.token,
+            &self_addr,
+            &stream.from,
+            to_balance
+        ));
+        EventBuilder::new()
+            .string("cancel_stream")
+            .address(&stream.from)
+            .address(&stream.to)
+            .number(stream_id)
+            .number(from_balance)
+            .number(to_balance)
+            .notify();
+        return true;
+    }
+    false
+}
+
+fn set_proxy(addr: Address) -> bool {
+    assert!(runtime::check_witness(&ADMIN));
+    database::put(KEY_PROXY, addr);
+    true
+}
+fn get_proxy() -> Option<Address> {
+    database::get::<_, Address>(KEY_PROXY)
+}
+
+fn get_stream(id: U128) -> Option<Stream> {
+    database::get::<_, Stream>(utils::gen_stream_key(id))
+}
+
+fn transfer(token: &Address, from: &Address, to: &Address, amount: U128) -> bool {
+    if token == &ONG_CONTRACT_ADDRESS {
+        return ong::transfer(from, to, amount);
+    } else if token == &ONT_CONTRACT_ADDRESS {
+        return ont::transfer(from, to, amount);
+    }
+    let token_new = get_migrate(token);
+    let ty = get_vm_ty(&token_new);
+    match ty {
+        NEO_VM => {
+            let mut sink = Sink::new(16);
+            sink.write(u128_to_neo_bytes(amount));
+            sink.write_neovm_address(to);
+            sink.write_neovm_address(from);
+            sink.write(83u8);
+            sink.write(193u8);
+            sink.write("transfer".to_string());
+            sink.write(103u8);
+            sink.write(&token_new);
+            let res = runtime::call_contract(&token_new, sink.bytes());
+            if let Some(data) = res {
+                let mut source = Source::new(&data);
+                return source.read().unwrap();
+            } else {
+                return false;
+            }
+        }
+        WASM_VM => {}
+        _ => panic!(""),
+    }
+    false
+}
+
+fn check_registered_token(token_addr: &Address) -> bool {
+    let tokens = get_registered_token();
+    if has_registered_token(&tokens, token_addr) {
+        return true;
+    }
+    false
+}
+
+fn update_stream_id(id: U128) {
+    database::put(KEY_STREAM_ID, id)
+}
+fn get_stream_id() -> U128 {
+    if let Some(id) = database::get::<_, U128>(KEY_STREAM_ID) {
+        return id;
+    }
+    1
+}
```

Comment:
`get` and `update` must be atomic to avoid bug.
```suggestion
fn get_next_stream_id() -> U128 {
  let id = database::get(KEY_STREAM_ID).unwrap_or(1);
  database::put(KEY_STREAM_ID, id+1);
  id
}
```